### PR TITLE
RISDEV-3766 Adjust db models and ports for new database tables

### PR DESCRIPTION
**Pull request by malte-laukoetter**
Todo:

* [x] saving of announcements
* [x] adjustments once the Norms Domain Object exists (see todo comments, also there are some changes in this PR to that class that need to be overwritten by the changes made to the norms domain object as they where only made here to be able to implement the other stuff here already)

### CODE REVIEW

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/NormMapper.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  /**
+   * Maps a {@link NormDto} to the domain {@link Norm}.
+   *
+   * @param normDto The input {@link NormDto} to be mapped.
+   * @return A new {@link Norm} mapped from the input {@link NormDto}.
+   */
+  public static Norm mapToDomain(final NormDto normDto) {
+    return Norm.builder().document(XmlMapper.toDocument(normDto.getXml())).build();
+    // TODO: (Malte Laukötter, 2024-04-15) should we assert/validate that the eli and guid are
+    // correct?
```
Comment (malte-laukoetter) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/NormMapper.java:23]: Any opinions?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/service/DBService.java [hunk 2]
```diff
@@ -100,6 +121,70 @@ public Optional<String> loadTargetLawXmlByEli(LoadTargetLawXmlPort.Command comma
...
+  }
+
+  @Override
+  public List<Announcement> loadAllAnnouncements() {
+    return announcementRepository.findAll().stream()
+        .map(AnnouncementMapper::mapToDomain)
+        .sorted(
+            Comparator.comparing(
+                    (Announcement announcement) ->
+                        announcement.getNorm().getPublicationDate().orElse(LocalDate.MIN))
```
Comment (malte-laukoetter) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/service/DBService.java:143]: i added this here as the existing loading for amending laws also does this sort, i'm a bit unsure if we should do this at this layer

File: backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java [hunk 3]
```diff
@@ -184,4 +187,27 @@ public static List<RomanNumeral> getReverseSortedValues() {
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    Norm norm = (Norm) o;
+    return document.isEqualNode(norm.document);
+  }
+
+  @Override
+  public int hashCode() {
```
Comment (malte-laukoetter) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java:200]: nothing really uses the hashCode but as we have overwritten the equals method need to fullfill the contract between the two methods. Document does not implement hashCode so it is comparing the object identities so we needed our own implementation. Just parsing it as html seemed like the best way to me, even if this means that we are now using some packages for this that we don't really want to have in the Domain layer (I added an exception for them for this specific class to the architecture test). 
If someone has a better idea how to do this I'm really interested :) (we want to be able to compare Norms using equals and that norms with the same xml are the same, this massively simplifies writing tests for them and probably also some other things in the future)

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/dto/AnnouncementDto.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder(toBuilder = true)
+@Entity
+@Table(name = "announcements")
+public class AnnouncementDto {
+
+  @Id @GeneratedValue private UUID id;
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/dto/AnnouncementDto.java:33]: Note for later: Should this be the GUID of the amending law?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/service/DBService.java [hunk 5]
```diff
@@ -100,6 +121,70 @@ public Optional<String> loadTargetLawXmlByEli(LoadTargetLawXmlPort.Command comma
...
+        .flatMap(
+            eli ->
+                announcementRepository
+                    .findByNormDtoEli(eli)
+                    .map(
+                        announcementDto -> {
+                          announcementDto.setReleasedByDocumentalistAt(
+                              announcement.getReleasedByDocumentalistAt());
+                          // It is not possible to change the norm associated with an announcement.
+                          // Therefore, we don't update that relationship.
```
Comment (malte-laukoetter) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/service/DBService.java:182]: this comment is wrong, the norm is also saved as CascadeType is ALL

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/AnnouncementMapper.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  /**
+   * Maps a domain {@link Announcement} to {@link AnnouncementDto}.
+   *
+   * @param announcement The input {@link Announcement} to be mapped.
+   * @return A new {@link AnnouncementDto} mapped from the input {@link Announcement}.
+   */
+  public static AnnouncementDto mapToDto(final Announcement announcement)
+      throws XPathExpressionException, TransformerException {
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/AnnouncementMapper.java:34]: Where does the TransformerException come from?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/NormMapper.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  /**
+   * Maps a domain {@link Norm} to {@link NormDto}.
+   *
+   * @param norm The input {@link Norm} to be mapped.
+   * @return A new {@link NormDto} mapped from the input {@link Norm}.
+   */
+  public static NormDto mapToDto(final Norm norm)
+      throws XPathExpressionException, TransformerException {
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/adapter/output/database/mapper/NormMapper.java:33]: Where's the TransformerException from?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/application/port/output/LoadAllAnnouncementsPort.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import de.bund.digitalservice.ris.norms.domain.entity.Announcement;
+import java.util.List;
+
+/** Port interface for loading all {@link Announcement}s from the storage. */
+public interface LoadAllAnnouncementsPort {
+
+  /**
+   * Loads all {@link Announcement}s available in the system.
+   *
+   * @return A {@link List} of {@link Announcement}, which may be empty if no Announcement are
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/application/port/output/LoadAllAnnouncementsPort.java:12]: "is found"

File: backend/src/main/java/de/bund/digitalservice/ris/norms/application/port/output/LoadNormPort.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  /**
+   * Loads a {@link Norm} based on the provided ELI specified in the command.
+   *
+   * @param command The command specifying the ELI to identify the norm to be loaded.
+   * @return An {@link Optional} containing the loaded {@link Norm} if found, or empty if not found.
+   */
+  Optional<Norm> loadNorm(final Command command);
+
+  /**
+   * A record representing the command for loading a norm. The command includes the ELI (European
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/application/port/output/LoadNormPort.java:22]: link?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Announcement.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package de.bund.digitalservice.ris.norms.domain.entity;
+
+import java.time.Instant;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+
+/**
+ * Class representing an article entity. This class is annotated with Lombok annotations for
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Announcement.java:9]: not an "article"

File: backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java [hunk 11]
```diff
@@ -184,4 +187,27 @@ public static List<RomanNumeral> getReverseSortedValues() {
           .toList();
     }
   }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (hannes-mk) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java:192]: maybe don't abbreviate?

File: backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java [hunk 12]
```diff
@@ -184,4 +187,27 @@ public static List<RomanNumeral> getReverseSortedValues() {
...
+
+    try {
+      new TransformerFactoryImpl()
+          .newTransformer()
+          .transform(new DOMSource(document), new StreamResult(writer));
+    } catch (TransformerException e) {
+      throw new XmlProcessingException(e.getMessage(), e);
+    }
+
+    return Objects.hash(writer.toString());
```
Comment (malte-laukoetter) [inline backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/Norm.java:211]: use XmlMapper

File: backend/src/test/java/de/bund/digitalservice/ris/norms/integration/adapter/output/database/DBServiceIntegrationTest.java [hunk 13]
```diff
@@ -514,4 +529,311 @@ void itUpdatesAmendingLaw() {
...
+    var norm = Norm.builder().document(XmlMapper.toDocument(xml)).build();
+    normRepository.save(NormMapper.mapToDto(norm));
+
+    // When
+    final Optional<Norm> normOptional =
+        dbService.loadNorm(
+            new LoadNormPort.Command("eli/bund/bgbl-1/1964/s593/1964-08-05/1/deu/regelungstext-1"));
+
+    // Then
+    assertThat(normOptional).isPresent().satisfies(normDb -> assertThat(normDb).contains(norm));
```
Comment (hannes-mk) [inline backend/src/test/java/de/bund/digitalservice/ris/norms/integration/adapter/output/database/DBServiceIntegrationTest.java:568]: can this be simplified?

**Pull request merged**